_posts/2015-08-19-Cocoa-Architecture:-Switchboard-Pattern.markdown

```diff
@@ -0,0 +1,104 @@
+---
+layout: post
+title: "Cocoa Architecture: ARSwitchboard"
+date: 2015-08-19
+comments: false
+categories: [ios, mobile, architecture, energy, eigen, eidolon]
+author: Orta Therox
+github-url: https://www.github.com/orta
+twitter-url: http://twitter.com/orta
+blog-url: http://orta.io
+---
+
+As a part of going through the design patterns we've found in the creation of the Artsy iOS apps, I'd like to introduce the Switchboard pattern. This evolved quite naturally out of [ARRouter](/blog/2015/08/15/Cocoa-Architecture:-Router-Pattern/) when applied to generating view controllers instead of API requests.
+
+<!-- more -->
+
+----------------
+
+## Where we started
+
+In what must be one of my best named commits, `b9ff28` aka *"CREATING THE VOID"* introduced an ARSwitchboard to Eigen. Aside from this being the commit where I could finally write `[ARVoidViewController theVoid]` it added support for tapping an Artwork on the home-screen and going to another view controller.
+
+We knew up-front that we needed to emulate the website's URL schemes, so we needed to come up with a way to support two ways of loading up a view controller. Here's what it looked like:
+
+```objc
+@implementation ARSwitchBoard
+
++ (void)setupRouter {
+    artsyHosts = [NSSet setWithObjects:@"art.sy", @"artsyapi.com", @"artsy.net", nil];
+
+    [RCRouter map:@"/artwork/:id" to:self with:@selector(loadArtworkFromURL:)];
+}
+
++ (void)navigateToURL:(NSURL * )url {
+    if([self isInternalURL:url] && [RCRouter canRespondToRoute:url.path]) {
+        [RCRouter dispatch:url.path];
+    } else {
+//        [self openInternalBrowser:url];
+    }
+}
+
++ (BOOL)isInternalURL:(NSURL * )url {
+    NSString * host = url.host;
+    if(host && [host hasPrefix:@"www"]) {
+        host = [host substringFromIndex:3];
+    }
+
+    return (host && [artsyHosts containsObject:host]);
+}
+
+#pragma mark -
+#pragma mark Artworks
+
++ (void)loadAttachmentCollection:(AttachmentCollection * )collection {
+    [[ARVoidViewController theVoid] loadArtworkViewWithAttachmentCollection:collection];
+}
+
++ (void)loadArtwork:(Artwork * )artwork {
+    AttachmentCollection * collection = [AttachmentCollection collectionWithItems:@[artwork] andIndex:0];
+    [self loadAttachmentCollection:collection];
+}
+
++ (void)loadArtworkFromURL:(NSDictionary * )options {
+    [[ARVoidViewController theVoid] loadArtworkViewWithID:options[@"id"]];
+}
+
+@end
+```
+
+It shows the pattern's humble origins quite well. It provides an API that any other view controller can call to handle pushing a view controller on to a stack somewhere, but not only that it can deal with being given an arbitrary URL and handle that too.
```

Comment:
I think this is missing some punctuation around ‘not only that’, otherwise it’s hard to read imo.